subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy

```diff
@@ -694,13 +695,60 @@ abstract class MakeGreen implements TransformAction<TransformParameters.None> {
         fails(":a:resolve")
 
         then:
-        // Documents existing behaviour. Should fail eagerly and with a better error message
-        failure.assertHasDescription("Execution failed for task ':a:resolve'.")
-        failure.assertHasCause("Execution failed for MakeGreen: ${file('b/build/b.jar')}.")
-        failure.assertHasCause("No service of type ${propertyType} available.")
+        if (configurationTime) {
+            failure.assertHasDescription("A problem occurred evaluating root project")
+            failure.assertHasCause("Cannot register artifact transform MakeGreen with parameters null")
+            failure.assertHasCause("Cannot use @InputArtifact annotation on property MakeGreen.getInput() of type ${typeName}. Allowed property types: java.io.File, org.gradle.api.provider.Provider.")
+        } else {
+            failure.assertHasDescription("Execution failed for task ':a:resolve'.")
+            failure.assertHasCause("Execution failed for MakeGreen: ${file('b/build/b.jar')}.")
+            failure.assertHasCause("Cannot use @InputArtifact annotation on property of type ${typeName}. Allowed property types: java.io.File, org.gradle.api.provider.Provider<org.gradle.api.file.FileSystemLocation>.")
+        }
+
+        where:
+        propertyType                                   | configurationTime
```

Comment:
How about a more descriptive name?

```suggestion
        propertyType                                   | expectFailureDuringConfigurationTime
```

Reply:
Why do we detect the problem at different times for different types?

Reply:
Since I can't I don't have a way to add the parameterized type `Provider<FileSystemLocation>` to the annotation on `InputArtifact`. Therefore, at registration time only the raw type is checked - `Provider`. At runtime, we have the parameterized type and can do the correct check.

Reply:
Somehow my comments have disappeared here. :/

Using `@ InjectionPointQualifier` is one way to connect `@InputArtifact` and `Provider< FileSystemLocation>`, but we don't necessarily have to use it if it doesn't retain enough information. We could record the connection via some other way where using a `TypeToken` is allowed. WDYT?

Reply:
I though about this, though it seems like an annotation is so nice to use. We could have a separate value on the annotation (something like `allowedProviders`) where we can capture the allowed types within a provider.

Do you think we should handle the allowed types in the `InjectAnnotationHandler` and provide a method there?

Reply:
So for option 1 we'd have something like this?

```java
public @interface InjectionPointQualifier {
    Class<?>[] supportedTypes() default {};
    Class<?>[] supportedProviderTypes() default {};
}
```

I think this could work. Putting the type on `InjectAnnotationHandler` is also good.

I think either could work as long as we don't lose interesting details. I think I like the `supportedProviderTypes` approach even better.

Reply:
Done.